File: UBB/Fourth Semester/Projecting and Programming Methods/Laboratory 06/Lab6JAVA/src/main/java/com/athleticscompetiton/service/Service.java
package com.athleticscompetiton.service;

import com.athleticscompetiton.domain.Child;
import com.athleticscompetiton.domain.Enrollment;
import com.athleticscompetiton.domain.Round;
import com.athleticscompetiton.domain.User;
import com.athleticscompetiton.repository.ChildrenInterface;
import com.athleticscompetiton.repository.EnrollmentsInterface;
import com.athleticscompetiton.repository.RoundsInterface;
import com.athleticscompetiton.repository.UsersInterface;
import utils.Pair;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.StreamSupport;

public class Service {
    private UsersInterface users_repo;
    private RoundsInterface rounds_repo;
    private ChildrenInterface children_repo;
    private EnrollmentsInterface enrollments_repo;

    public Service(UsersInterface users, RoundsInterface rounds, ChildrenInterface children, EnrollmentsInterface enrollments)
    {
        this.users_repo = users;
        this.rounds_repo = rounds;
        this.children_repo = children;
        this.enrollments_repo = enrollments;
    }

    public User findUser(String username, String password)
    {
        return users_repo.findUser(username, password);
    }

    public long getNrChildrenForRound(String name)
    {
        Round round = rounds_repo.findOneName(name);
        return StreamSupport.stream(enrollments_repo.findChildrenForRound(round.getId()).spliterator(),false).count();
    }

    public Iterable<Round> findAllRounds()
    {
        return rounds_repo.findAll();
    }

    public Enrollment FindEnrollmentByChildAndRound(String firstname, String lastname, String roundName)
    {
        Child child = children_repo.findOneName(firstname, lastname);
        Round round = rounds_repo.findOneName(roundName);
        return enrollments_repo.findOne(new Pair(child.getId(), round.getId()));
    }

    public void SaveChild(String firstname, String lastname, int age)
    {
        Child child = new Child(firstname, lastname, age);
        children_repo.save(child);
    }

    public Iterable<Child> findChildrenForRound(String roundName)
    {
        return enrollments_repo.findChildrenForRound(rounds_repo.findOneName(roundName).getId());
    }

    public Child FindChildByName(String firstname, String lastname)
    {
        return children_repo.findOneName(firstname, lastname);
    }

    public Iterable<Child> FindChildrenByAgeGroup(String ageGroup)
    {
        Iterable<Child> children = children_repo.findAll();
        List<Child> e = new ArrayList<Child>();
        for(Child child :  children)
            if (child.getAgeGroup() == ageGroup)
                 e.add(child);
        return e;
    }

    public Enrollment SaveEnrollment(String firstname, String lastname, int age, String roundName)
    {
        Child child = children_repo.findOneName(firstname, lastname);
        Round round = rounds_repo.findOneName(roundName);
        Enrollment enrollment = new Enrollment(child, round);
        return enrollments_repo.save(enrollment);
    }

    public long FindNumberOfEnrollmentsForChild(String firstname, String lastname)
    {
        Child child = children_repo.findOneName(firstname, lastname);
        Iterable<Round> roundsForChild = enrollments_repo.findRoundsForChild(child.getId());
        return StreamSupport.stream(roundsForChild.spliterator(),false).count();
    }
}

File: UBB/Fourth Semester/Projecting and Programming Methods/Laboratory 05/Lab5C#/AthleticsCompetition/service/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AthleticsCompetition.domain;
using AthleticsCompetition.repository;
using AthleticsCompetition.utils;

namespace AthleticsCompetition.service
{
    public class Service : Observable
    {

        private IUsers users_repo;
        private IRounds rounds_repo;
        private IChildren children_repo;
        private IEnrollments enrollments_repo;
        private IList<Observer> observers = new List<Observer>();

        public Service(IUsers users, IRounds rounds, IChildren children, IEnrollments enrollments)
        {
            this.users_repo = users;
            this.rounds_repo = rounds;
            this.children_repo = children;
            this.enrollments_repo = enrollments;
        }

        public User findUser(String username, String password)
        {
            return users_repo.FindUser(username, password);
        }

        internal object getNrChildrenForRound(string name)
        {
            Round round = rounds_repo.FindOneName(name);
            return enrollments_repo.FindChildrenForRound(round.Id).Count;
        }

        public IEnumerable<Round> findAllRounds()
        {
            return rounds_repo.FindAll();
        }

        public override void AddObserver(Observer o)
        {
            throw new NotImplementedException();
        }

        public override void NotifyObservers()
        {
            throw new NotImplementedException();
        }

        public Enrollment FindEnrollmentByChildAndRound(string firstname, string lastname, string roundName)
        {
            Child child = children_repo.FindOneName(firstname, lastname);
            Round round = rounds_repo.FindOneName(roundName);
            return enrollments_repo.FindOne(Tuple.Create(child.Id, round.Id));
        }

        internal void SaveChild(string firstname, string lastname, int age)
        {
            Child child = new Child(firstname, lastname, age);
            children_repo.Save(child);
        }

        public IEnumerable<Child> findChildrenForRound(String roundName)
        {
            return enrollments_repo.FindChildrenForRound(rounds_repo.FindOneName(roundName).Id);
        }

        public Child FindChildByName(string firstname, string lastname)
        {
            return children_repo.FindOneName(firstname, lastname);
        }

        public IEnumerable<Child> FindChildrenByAgeGroup(string ageGroup)
        {
            List<Child> children = children_repo.FindAll();
            List<Child> e = new List<Child>();
            foreach (Child child in children)
            {
                if (child.GetAgeGroup() == ageGroup)
                    e.Add(child);
            }
            return e;
        }

        public Enrollment SaveEnrollment(string firstname, string lastname, int age, string roundName)
        {
            Child child = children_repo.FindOneName(firstname, lastname);
            Round round = rounds_repo.FindOneName(roundName);
            Enrollment enrollment = new Enrollment(child, round);
            return enrollments_repo.Save(enrollment);
        }

        public int FindNumberOfEnrollmentsForChild(string firstname, string lastname)
        {
            Child child = children_repo.FindOneName(firstname, lastname);
            List<Round> roundsForChild = enrollments_repo.FindRoundsForChild(child.Id);
            return roundsForChild.Count;
        }
    }
}
